atexit: move hook before precompile output

vtjnash opened the pull request:
To show how this is used, this updates the profile_printing_listener background job to use this mechanism.

vtjnash committed:
atexit: move hook before precompile output

To show how this is used, this updates the profile_printing_listener
background job to use this mechanism.

Also ensure mktemp and mktempdir cleanup even on unexpected exit:
Previously this was relying on all Tasks running to completion, which is
not a good assumption. Add the atexit hook, so that even if they do not
run to completion, they still get cleaned up.

JeffBezanson reviewed (approved):
I think this is the right thing to do. But noting that we should properly expose `_wait` and `_trywait` since there are clearly good use cases for them.

vtjnash commented:
Looks like it currently also breaks a file.jl test that I will need to look into

vtjnash committed:
restore deprecated function for Pkg tests

vtjnash committed:
improve "deadlock avoidance" of unexpectedly abrubt exits

vtjnash committed:
test: fix race in atexit test

Exposed by the presence of any atexit hooks in the new process, such as added in this PR.

johnnychen94 commented:
Curiously, will this be backported to 1.10?
